ZENKO-3023 phase 1 object squad release process

tcarmet opened the pull request:
# Phase 1 Object Squad Release

The goal of this PR is to document our release process.

It will be written here so that we can review together, approve it and act uppon those words.

The goal of the first phase is to disconnect our repository, if approved we apply this words as much as we can. (we can consider to trash this theory if we find a better one, no harm feelings ;) I'm open to any other suggestions)

PS: @wabernatScality before you go wild on my english mistakes, please allow some time to actually validate the content, I'll give you a go ;)

miniscruff commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,52 @@
+
+## Components
+
+We have multiple libraries, API, tools that we code in house. Let’s treat them as any open source component we would like to use in a project.
```
Small formatting comment I am mentioning as it helps, me at least, review markdown. I put each sentence on a new line as markdown will merge them all into one paragraph. But for reviewing and detailing it's nice to get that split. Also helps when leaving comments as you can only make line comments and not a word or phrase comments.

tcarmet replied:
True, it doesn't change the output, I'll do that to see how it looks (code wise).

tcarmet replied:
done

miniscruff commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,52 @@
+We have multiple libraries, API, tools that we code in house. Let’s treat them as any open source component we would like to use in a project.
+
+Each component should contain:
+* Proper code
```
I would remove this as it ignores documentation repos and config repos ( eslint ), and it is on top people might over prioritize it and skip the others.

tcarmet replied:
Ok
```suggestion
```

miniscruff commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,52 @@
+
+We have multiple libraries, API, tools that we code in house. Let’s treat them as any open source component we would like to use in a project.
+
+Each component should contain:
```
I know you didn't want grammar but I would replace should with must, see https://tools.ietf.org/html/rfc2119

tcarmet replied:
HAHAHA fair enough let's change it
```suggestion
Each component must contain:
```

miniscruff commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,52 @@
+
+Selecting a component should be as easy and understandable as picking one from the open source world. Either from it’s proper language registry. (bumping the version on the package.json file per example for nodejs projects) or pulling a docker image from a registry. Therefore it is the component responsibility to make those available.
+
+Each component release is made by the project leader with the help of the CI chapter, advertised internally and externally if needed. Release notes are provided with each release.
```
I feel project leader should be component leader, or just repository owner(s). Including a CODEOWNERS file per repo could also be another requirement to further distinguish this difference.

tcarmet replied:
So project leader is just the world we are already use to, here at Scality (it's also how the config is named in the bot bert-e but that's just a detail). I don't mind changing the wording to repository owners. As long as it make sense to all of us.

cc: @rahulreddy are you fine with that wording?

The codeowner is also an interesting tool to use. I'm thinking about creating a ticket for it, instead of documenting and tackling it right now, so that there's not too much to do at once, but I could be wrong, what's your thought?  @miniscruff

miniscruff replied:
Code owner gives more control like per folder and maybe per branch, probably not required to add it just yet. But moving forward it may be nice, maybe as a replacement for eve settings it just uses code owners.

tcarmet replied:
So just as an info to ensure that we're all on the same page the settings are for Bert-E, not for eve. Bert-E takes care of gatekeeping the pull requests, and Eve does CI.

But yes, that could be an interesting thing to do, we didn't at first, because we were supporting both Bitbucket and Github and they do not behave the same way, now that we only have to interact with GitHub it makes things easier to implement.

As per the per folder settings, yes I see that this can be very useful especially for docs related stuff where we will need to ensure that William had a look at it per example. But as you're mentioning, let's do baby steps, using `CODEOWNERS` with the whole repo for now, and then we move forward to specific folders. 

We can, therefore add a mention to specify that each repository have a `CODEOWNERS` file up to date. How does that sound?

miniscruff commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,52 @@
+
+Project release should be as small as possible. Per example just bumping the version of a component can unlock feature or bugfixes. Therefore provides value. As long as we keep shipping releases we are unlocking customers by small packs and we get the feedback we need along the way.
+
+It is the project responsibility to make all required packages for installation or installation instructions available. 
```
Just to be clear here, you are just saying the documentation is available and not that we have to write it specifically in the project? That is, zenko doesn't need cert-manager docs and can instead link to the component?

tcarmet replied:
Correct, I don't think we need to  re document what's already out there, IMHO. But then in some situation it might be worth writing a word about it.
The idea is mainly to say that the documentation of Zenko resides inside the repo and contain the required information to how to install it and use it. For cert-manager or any other component if we point to the doc of it, it's fine. At the product level though, it could be a different story, but that's not up to the project.

miniscruff reviewed (approved):
Even if we need to adjust later, I think there is more than enough improvements to get started here.

tmacro commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+# Disconecting Releases
+
+From the moment we are in a multi repository environment.
+We need to disconect the releases of each of our repositories.
+Define how we release them depending on the role of the repository itself, and then how do we combine them all together.
```
```suggestion
We need to disconnect the releases of each of our repositories, define how we release them depending on the role of the repository itself, and how we combine them all together.
```

tmacro commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+Selecting a component should be as easy and understandable as picking one from the open source world.
+Either from it’s proper language registry like bumping the version on the package.json file per example for nodejs projects or pulling a docker image from a registry.
+Therefore it is the component responsibility to make those available.
```
```suggestion
Therefore it is the component's responsibility to make those available.
```

tcarmet commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,61 @@
+
+## Product
+
+The product repository are what is shipped to the customer.
```
```suggestion
The product repositories are what is shipped to the customer.
```

terrydervaux commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,61 @@
+* Assemble *projects* and *components* with their respective version.
+* Proper build chain to output the installation code and packages.
+* Integration tests
+* Documentation
```
Maybe a little bit more details.
```suggestion
* Documentation to explain the services provided by the product, how to install, configure, monitor, and solve product issues.
```

terrydervaux commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,61 @@
+
+## Components
+
+We have multiple libraries, API, tools that we code in house.
```
Can you add a few examples like vault, cloudserver, arsenal...?

tcarmet replied:
How does this sound?
```suggestion
We have multiple libraries, API, tools that we code in house.
Per example, Arsenal, Cloudserver, Backbeat and so on.
```

terrydervaux reviewed (approved):
I added a few suggestions, otherwise LGTM.

tcarmet commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+Each component must contain:
+* Unit and functional tests to ensure that the component itself work as it should.
+* Documentation
```
good point.

tcarmet commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,52 @@
+* Unit and functional tests to ensure that the component itself work as it should.
+* Documentation
+* Release notes
+
```
I already mentioned test above, do you mind if I modify it to:
```suggestion
* Build chain to build and publish packages
```

terrydervaux replied:
I'd prefer that you keep "test" on the list.

tcarmet commented:
@wabernatScality I believe you can start working your magic here :mage:

anurag4DSB commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+
+Here are the subjects that we are trying to accomplish by applying this release process:
+* Move forward to a simpler and conventional release workflow.
+* Faster development process. improving code complete to GA times and end2end tests feedbacks.
```
```suggestion
* Faster development process. Improving code complete to GA times and end2end tests feedbacks.
```

anurag4DSB reviewed (commented):
A small comment and question:

- How does zenko bring in the release notes from each individual component? Will it bring them? If not, it should, right?

tcarmet commented:
> A small comment and a question:
> 
>     * How does zenko bring in the release notes from each individual component? Will it bring them? If not, it should, right?

Ah HA! Very good question, here are my thoughts towards that idea. Maybe we should write it down.

I don't think we should bring the release note from each individual component. We don't bring the release of mongo db or kafka when we make an upgrade of those, so why should we bring the ones of the component (kafka and mongo are components of zenko just like cloudserver)?

We may however still write a release note when we upgrade a component (could be kafka, mongodb or our own) but it should translate to something that speaks to Zenko. The release note of the component itself will be too specific to the component and may not speak to a Zenko user. And that is why I think it should be rewritten to something more Zenko Friendly.

To add more point towards that argument, say we update Cloudserver. Cloudserver new releases contain a new function that can do cool things, but we may or may not be using it in Zenko, so we shouldn't write about it. What we will write, is that we corrected this bug add a little bit of explanation of it, and the commit content is a bump of cloudserver. Same works for features, I'm gonna write a release note that talks about this new API that I'm adding, the content of my commit may be bumps in cloudserver version and vault's version, but I'm not gonna copy the full changelog of those, I'm just going to say, we've added this new API that does this. At the end, I think this is what the Zenko user will be interested in. 


I'll leave in cc: @terrydervaux, @wabernatScality, @rahulreddy  and @miniscruff to see if they have something against this "concept"

anurag4DSB commented:
> ery good question, here are my thoughts towards that idea. Maybe we should write it down.

I am thinking about a part of release notes. Specifically, known issues and fixed issues. So may be we need them from our own components. But that can be manual too. In any case, needs to be clear :). I like your plan in this PR so approving.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+## Project
+
+Project repository, per example Zenko.
```
Fragment. I would move this down or delete it entirely. The next sentence is much more relevant and direct.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+Project repository, per example Zenko.
+
+The role of a project repository is to:
```
The purpose of a project repository is to:

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+Project repository, per example Zenko.
+
+The role of a project repository is to:
+* Select *components* with their respective version, either developed in house or open source one.
```
* Select components (developed in-house or from open source projects) by version.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+The role of a project repository is to:
+* Select *components* with their respective version, either developed in house or open source one.
+* Creating the proper build chain to output the desired installation code and packages.
```
* Create a build chain that outputs the desired installation code and packages.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+The role of a project repository is to:
+* Select *components* with their respective version, either developed in house or open source one.
+* Creating the proper build chain to output the desired installation code and packages.
+* Integration tests to ensure that all components selected work well together.
```
* Trigger integration tests that ensure all selected components interoperate effectively.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Select *components* with their respective version, either developed in house or open source one.
+* Creating the proper build chain to output the desired installation code and packages.
+* Integration tests to ensure that all components selected work well together.
+* Documentation to explain the services provided by the project, how to contribute, build, test, configure, run, and monitor the project.
```
* Document the project's services, and how to build, test, configure, run, monitor, and contribute to the project.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Creating the proper build chain to output the desired installation code and packages.
+* Integration tests to ensure that all components selected work well together.
+* Documentation to explain the services provided by the project, how to contribute, build, test, configure, run, and monitor the project.
+* Release notes
```
* Offer release notes that document changes in the product's function and features.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Documentation to explain the services provided by the project, how to contribute, build, test, configure, run, and monitor the project.
+* Release notes
+
+Project release should be as small as possible.
```
Project releases should be kept as small as possible. For example, bumping the version of a component might unlock features or bugfixes. If we keep shipping releases, we can advance customers a few at a time and receive needed feedback along the way.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+Therefore provides value.
+As long as we keep shipping releases we are unlocking customers by small packs and we get the feedback we need along the way.
+
+It is the project responsibility to make all required packages for installation or installation instructions available. 
```
The responsibility for providing instructions and all required packages for installation is at the *project* level, not the *product* level.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+It is the project responsibility to make all required packages for installation or installation instructions available. 
+
+The releases are made by the Project Owner. 
```
Release readiness is determined by the project owner.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+## Product
+
+The product repositories are what is shipped to the customer.
```
Product repositories contain the code and documentation that are shipped to the customer.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+The product repositories are what is shipped to the customer.
+
+The role of a product repository is to:
```
The purpose of a product repository is to:

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+The product repositories are what is shipped to the customer.
+
+The role of a product repository is to:
+* Assemble *projects* and *components* with their respective version.
```
You need to unpack this (The bad word is "respective".).

* Assemble projects and components under a common _product_ version?

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+The role of a product repository is to:
+* Assemble *projects* and *components* with their respective version.
+* Proper build chain to output the installation code and packages.
```
* Create a build chain that outputs the desired installation code and packages.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+The role of a product repository is to:
+* Assemble *projects* and *components* with their respective version.
+* Proper build chain to output the installation code and packages.
+* Integration tests
```
* Trigger integration tests 
[that ensure all selected components interoperate effectively.]?

tcarmet replied:
In this case, at the product level it mostly ensure all *projects* (and some components) interoperate effectively

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Assemble *projects* and *components* with their respective version.
+* Proper build chain to output the installation code and packages.
+* Integration tests
+* Documentation to explain the services provided by the product, how to install, configure, monitor, and solve product issues.
```
* Document the product's services; how to install, configure, run, and monitor the product; and how to solve product issues.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Proper build chain to output the installation code and packages.
+* Integration tests
+* Documentation to explain the services provided by the product, how to install, configure, monitor, and solve product issues.
+* Release notes
```
* Offer release notes that document changes in the project's function and features.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Documentation to explain the services provided by the product, how to install, configure, monitor, and solve product issues.
+* Release notes
+
+Releases of the product are not defined inside the object squad and they are more based on customer needs.
```
Product releases are driven by customer needs and are not defined within the object squad.

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Release notes
+
+Releases of the product are not defined inside the object squad and they are more based on customer needs.
+However it is the object squad responsability to make project releases available for the product and help in the integration work.
```
The object squad, however, is responsible for making project releases available to the product and for helping integrate the projects into the product.

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+# Object Squad Release process
```
Process

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+# Object Squad Release process
+
+This is the official documentation of how we will handle releases for the object squad.
```
documentation for how the Object Squad will handle releases.

tcarmet replied:
normal that we're repeating documentation twice?

tcarmet replied:
```suggestion
This is the official documentation for how the Object Squad will handle releases.
```

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+This is the official documentation of how we will handle releases for the object squad.
+
+
+Here are the subjects that we are trying to accomplish by applying this release process:
```
We expect this release process will:

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+
+
+Here are the subjects that we are trying to accomplish by applying this release process:
+* Move forward to a simpler and conventional release workflow.
```
* Give us a simpler, more conventional release workflow.

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+
+Here are the subjects that we are trying to accomplish by applying this release process:
+* Move forward to a simpler and conventional release workflow.
+* Faster development process. Improving code complete to GA times and end2end tests feedbacks.
```
* Speed our development, improving code-complete-to-GA time and end-to-end test feedback.

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+Here are the subjects that we are trying to accomplish by applying this release process:
+* Move forward to a simpler and conventional release workflow.
+* Faster development process. Improving code complete to GA times and end2end tests feedbacks.
+* We need to release more and smaller.
```
* Enable us to release more and smaller.

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+* Move forward to a simpler and conventional release workflow.
+* Faster development process. Improving code complete to GA times and end2end tests feedbacks.
+* We need to release more and smaller.
+* Untie every project we have and treat them indepently.
```
* Untangle our projects and give them independence.

wabernatScality commented on docs/release/index.md:
```diff
@@ -0,0 +1,13 @@
+* Faster development process. Improving code complete to GA times and end2end tests feedbacks.
+* We need to release more and smaller.
+* Untie every project we have and treat them indepently.
+* Remove the stress and intensity around release periods.
```
* Diminish the stress around releases.

wabernatScality reviewed (commented):
@tcarmet: Here you go!

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+## Components
+
+We have multiple libraries, API, tools that we code in house.
+Per example, Arsenal, Cloudserver, Backbeat and so on.
```
```suggestion
, for example, Arsenal, CloudServer, and Backbeat.
```

wabernatScality reviewed (commented):
@tcarmet: Thank @miniscruff for educating me as to the correct use of the suggestion feature. I hope this makes this wall of corrections a little less awful!

ilkescality reviewed (approved):
current suggestions look good, aside from it LGTM!

tcarmet commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+
+Each component release is made by the project leader with the help of the CI chapter, advertised internally and externally if needed.
+Release notes are provided with each release.
+
```
done

tcarmet commented:
> @tcarmet: Thank @miniscruff for educating me as to the correct use of the suggestion feature. I hope this makes this wall of corrections a little less awful!

OMG yes, that made my life easier.

Also huge thanks! took me a while to read them all, so I can imagine the work on your side to make them. Thanks a lot. It is very appreciated.

Took almost all of them into account, left some answer and questions along the way, tell me what you think @wabernatScality !

Thanks again :')

wabernatScality commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+# Disconecting Releases
+
+From the moment we are in a multi repository environment.
+We need to disconnect the releases of each of our repositories, define how we release them depending on the role of the repository itself, and how we combine them all together.
```
We must redefine how our products are connected to their component repositories, taking how we intend to release products into consideration.

tcarmet replied:
```suggestion
We must redefine how our products are connected to their component repositories, taking how we intend to release products into consideration.
```

tcarmet commented on docs/release/disconnecting.md:
```diff
@@ -0,0 +1,62 @@
+* Documentation to explain the services provided by the component, how to contribute, build, test, configure, run, use, and potentially monitor the component.
+* Release notes
+* Build chain to build, test, and publish packages
+They will then have their own releases independently from the state of the product itself.
```
```suggestion
Each repository will thus have its own release cycle that is independent of the release cycle of the product itself.
```

wabernatScality reviewed (approved):
I couldn't comment in line, so I'll do it here: Apropos the question of release notes, we should, I think, document changes in the components and pass them through iff they are implemented in the product. In other words, if Bucket Notification were active for S3C but not Zenko, we would document the API once in CloudServer, then either include or exclude the note from Zenko/S3C release notes. 

I approve...
